Rare interests core project

BoominHutJr opened the pull request:
Finished basic implementation, working on test cases.

erikqu commented on pipeline/matching/generators/rare_interests.py:
```diff
@@ -1,7 +1,10 @@
-from typing import Iterator
+# used to count occurences in a list
```
Remove misc. comments unless clever logic (applies everywhere)

erikqu commented on pipeline/matching/generators/rare_interests.py:
```diff
@@ -14,6 +17,78 @@ def __init__(self, max_frequency: float = 0):
+            ]
+            yield from m
+
+        usrList = inp.users
```
feel free to rename the input var name in this method

erikqu commented on pipeline/matching/generators/rare_interests.py:
```diff
@@ -14,6 +17,78 @@ def __init__(self, max_frequency: float = 0):
+        rarestInterest = " "
+        rarityScore = 0.0
+        matchNames = set("None")
+        m = []
```
Lets name this something meaningful

erikqu commented on pipeline/matching/generators/rare_interests.py:
```diff
@@ -14,6 +17,78 @@ def __init__(self, max_frequency: float = 0):
+                users=matchNames,
+                metadata=MatchMetadata(
+                    generator="rareInterestsGenerator",
+                    score=round(rarityScore, 2),
```
Nice, rounding this was a good idea

BoominHutJr commented:
- added tests to cover cases  in which there are 1 rare interest, multiple rare interests, and all interests are equally rare
- fixed null input check
- fixed rare interest list reflecting rare interests of the community rather than individual matches

BoominHutJr commented:
- added test for scenario of a community that has no common interests

erikqu commented on pipeline/matching/generators/rare_interests_test.py:
```diff
@@ -37,13 +37,143 @@ def test_example():
+            User(uid="4", displayName="D", interests=[swimming, cooking]),
+        ],
+        recent_matches=[],
+        interests=[hiking, swimming, reading],
```
Should we have cooking here?

BoominHutJr replied:
yes, thanks for catching that

bridgettorres commented:
LGTM!

bridgettorres commented on pipeline/matching/generators/rare_interests.py:
```diff
@@ -13,7 +16,80 @@ def __init__(self, max_frequency: float = 0):
+                ),
+            )
+        ]
+        yield from resultingMatch
```
Everything looks clean and easy to read, LGTM!

bridgettorres commented on pipeline/matching/generators/rare_interests_test.py:
```diff
@@ -37,13 +37,143 @@ def test_example():
+        )
+    ]
+    assert len(actual) == 1
+    assert actual == expected
```
Good test cases, clear names!

erikqu commented on pipeline/matching/generators/rare_interests.py:
```diff
@@ -13,7 +16,80 @@ def __init__(self, max_frequency: float = 0):
+                    rareInterests_.append(rarestInterest)
+                rarityScore = float(rarityScoreDict[rarestInterest])
+
+        # convert to set and back to list to remove duplicates
```
Lets remove this comment

erikqu commented on pipeline/matching/generators/rare_interests.py:
```diff
@@ -13,7 +16,80 @@ def __init__(self, max_frequency: float = 0):
+            )
+            for commonInterest in commonInterestList:
+                rarestInterest = min(
+                    rarityScoreDict.items(), key=lambda x: x[1]
```
This could be perceived as clever, maybe a comment here would be nice
